pkg/common/policy/policy.go

```diff
@@ -0,0 +1,170 @@
+	Caller string `json:"caller"`
+
+	// FullMethod is the fully-qualified name of the proto rpc service method.
+	FullMethod string `json:"full_method"`
+
+	// Req represents data received from the request body. It MUST be a
+	// protobuf request object with fields that are serializable as JSON,
+	// since they will be used in policy definitions.
+	Req interface{} `json:"req"`
+}
+
+type Result struct {
+	Allow             bool `json:"allow"`
+	AllowIfAdmin      bool `json:"allow_if_admin"`
+	AllowIfLocal      bool `json:"allow_if_local"`
+	AllowIfDownstream bool `json:"allow_if_downstream"`
+	AllowIfAgent      bool `json:"allow_if_agent"`
+}
+
+// NewEngine returns a new policy engine. Or nil if no
+// config is provided.
+func NewEngine(cfg *EngineConfig) (*Engine, error) {
+	if cfg == nil || cfg.FileProvider == nil {
+		return nil, nil
+	}
+
+	module, err := ioutil.ReadFile(cfg.FileProvider.PolicyPath)
+	if err != nil {
+		return nil, err
+	}
+	storefile, err := os.Open(cfg.FileProvider.PermissionsPath)
+	if err != nil {
+		return nil, err
+	}
+	defer storefile.Close()
+	store := inmem.NewFromReader(storefile)
+
+	return NewEngineFromRego(string(module), store)
+}
+
+// DefaultAuthPolicy returns the default policy engine
+func DefaultAuthPolicy() (*Engine, error) {
+	var json map[string]interface{}
+	err := util.UnmarshalJSON([]byte(defaultPermissionData), &json)
+	if err != nil {
+		return nil, err
+	}
+	store := inmem.NewFromObject(json)
+
+	return NewEngineFromRego(defaultPolicyRego, store)
+}
+
+// NewEngineFromRego is a helper to create the Engine object
+func NewEngineFromRego(regoPolicy string, dataStore storage.Store) (*Engine, error) {
+	rego := rego.New(
+		rego.Query("data.spire.result"),
+		rego.Package(`spire`),
+		rego.Module("spire.rego", regoPolicy),
+		rego.Store(dataStore),
+	)
+	pr, err := rego.PartialResult(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	e := &Engine{
+		rego: pr,
+	}
+
+	// Test policy with some simple calls to ensure that the
+	// policy can be evaluated properly.
+	if err = e.testPolicy(); err != nil {
+		return nil, err
+	}
+
+	return e, nil
+}
+
+func (e *Engine) testPolicy() error {
+	ctx := context.Background()
+	for _, i := range sampleInputs {
+		var inp Input
+		err := json.Unmarshal([]byte(i), &inp)
+		if err != nil {
+			return err
+		}
+
+		_, err = e.Eval(ctx, inp)
+		if err != nil {
+			return errors.New("policy initialization: issue testing configured policy")
+		}
+	}
+	return nil
+}
+
+// Eval determines whether access should be allowed on a resource.
+func (e *Engine) Eval(ctx context.Context, input Input) (result Result, err error) {
+	rs, err := e.rego.Rego(rego.Input(input)).Eval(ctx)
+	if err != nil {
+		return result, err
+	}
+
+	// TODO(tjulian): figure this out
+	if len(rs) == 0 || len(rs[0].Expressions) == 0 {
+		return result, errors.New("policy: no matching policies found")
+	}
+
+	exp := rs[0].Expressions[0]
+	resultMap := exp.Value.(map[string]interface{})
```

Comment:
Is the value guaranteed to be of this type? If not, we should use the comma-ok pattern on the type coercion to prevent a panic and return an error instead.